### PR TITLE
Adding GRUv3 support.

### PR DESCRIPTION
**Pull request by mkschleg**
As per the starting discussion in #1671, we should provide support for variations on the GRU and LSTM cell. 

In this PR, I added support for the GRU found in v3 of the [original GRU paper](https://arxiv.org/abs/1406.1078). Current support in Flux is for v1 only. [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/layers/GRU) supports several variations, with this as one of the variations. 

While the feature is added and usable in this PR, this is only a first pass at a design and could use further iterations. Some questions I have:
- Should we have new types for each variation of these cells? (another possibility is through parametric options)
- Should we have a shared constructor similar to Tensorflow/Pytorch? (it might make sense to rename the current GRU to GRUv1 if we want to do this).

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable
- [ ] API changes require approval from a committer (different from the author, if applicable)

### CODE REVIEW

**Comment by mkschleg** [comment]
I added tests to `test/layers/recurrent.jl`. I'm not sure there are more tests I should add? Also added doc strings for the new method and updated the doc string for the original version to clarify it is v1 of the arxiv paper.

File: src/layers/recurrent.jl [hunk 1]
```diff
@@ -233,6 +234,62 @@ function Base.getproperty(m::GRUCell, sym::Symbol)
...
+
+struct GRUv3Cell{A,V,S}
+  Wi::A
+  Wh::A
+  b::V
+  Wh_h̃::A
+  state0::S
+end
+
+GRUv3Cell(in, out; init = glorot_uniform, initb = zeros32, init_state = zeros32) =
```
Comment (DhairyaLGandhi) [inline src/layers/recurrent.jl:248]: Needs an activation

File: src/layers/recurrent.jl [hunk 2]
```diff
@@ -233,6 +234,62 @@ function Base.getproperty(m::GRUCell, sym::Symbol)
...
+    GRUv3(in::Integer, out::Integer)
+
+[Gated Recurrent Unit](https://arxiv.org/abs/1406.1078) layer. Behaves like an
+RNN but generally exhibits a longer memory span over sequences. This implements
+the version proposed in v3 of the above arxiv paper.
+
+See [this article](https://colah.github.io/posts/2015-08-Understanding-LSTMs/)
+for a good overview of the internals.
+"""
+GRUv3(a...; ka...) = Recur(GRUv3Cell(a...; ka...))
```
Comment (DhairyaLGandhi) [inline src/layers/recurrent.jl:278]: How do the versions differ api wise? Does this need any extra terms?

**Review by DhairyaLGandhi — COMMENTED** [review]
Thanks for the contribution! Looking forward to it!

I've left a couple of starter comments. Could you please also add CUDA tests?

File: src/layers/recurrent.jl (hunk 2 above)
Comment (mkschleg) [reply src/layers/recurrent.jl:278]: We shouldn't. The main difference is the `Wh` matrix needs to be split so we can apply the reset vector appropriately.  But this doesn't require any extra parameters for the constructor.

**Comment by CarloLucibello** [comment]
there seems to be a lot of code duplication, probably adding a symbol parameter to the type, `GRU{..., Mode}`, and only dispatch the call method on that seems a better alternative

**Comment by mkschleg** [comment]
Right. That is definitely the question I had. The conversation was derailed a bit in #1671, as we started discussing CuDNN support. 

If we were to do the parametric option version we would have to figure out how to do the operation in the line I tagged you in with the current struct layout (it is [this line](https://github.com/FluxML/Flux.jl/blob/2cf128816e73a704e31bc420cfaf57fdb017b0d1/src/layers/recurrent.jl#L257). I'm not sure how that would work tbh without extra unnecessary operations.

File: src/layers/recurrent.jl [hunk 3]
```diff
@@ -233,6 +234,62 @@ function Base.getproperty(m::GRUCell, sym::Symbol)
...
+GRUv3Cell(in, out; init = glorot_uniform, initb = zeros32, init_state = zeros32) =
+    GRUv3Cell(init(out * 3, in), init(out * 2, out), initb(out * 3), 
+              init(out, out), init_state(out,1))
+
+function (m::GRUv3Cell{A,V,<:AbstractMatrix{T}})(h, x::Union{AbstractVecOrMat{T},OneHotArray}) where {A,V,T}
+  b, o = m.b, size(h, 1)
+  gx, gh = m.Wi*x, m.Wh*h
+  r = σ.(gate(gx, o, 1) .+ gate(gh, o, 1) .+ gate(b, o, 1))
+  z = σ.(gate(gx, o, 2) .+ gate(gh, o, 2) .+ gate(b, o, 2))
+  h̃ = tanh.(gate(gx, o, 3) .+ (m.Wh_h̃ * (r .* h)) .+ gate(b, o, 3))
```
Comment (mkschleg) [inline src/layers/recurrent.jl:257]: @CarloLucibello This line.

**Comment by mkschleg** [comment]
@DhairyaLGandhi cuda tests added.

**Comment by DhairyaLGandhi** [comment]
It might make sense to have a function that is a no-op for the regular GRU layer (without changing the struct) and splits the array for v3.

**Comment by mkschleg** [comment]
I remember having issues with views + zygote awhile ago. Is that still the case? (I've not tried since v0.10.x). If we can't use views the split function would create a larger memory footprint, no? I think @darsnack had some thoughts on this in terms of clarity of API.

**Comment by DhairyaLGandhi** [comment]
Should be the same memory wise

File: src/layers/recurrent.jl (hunk 1 above)
Comment (darsnack) [reply src/layers/recurrent.jl:248]: AFAIK the activations in LSTM/GRU are very specifically chosen. That's why they are currently not options, and we should probably keep that consistent.

**Comment by mkschleg** [comment]
I think I misunderstood what you suggested. This would be in the constructor, right? Then I agree memory would be the same.

We would have to add an extra parametric type to separate Wi and Wh (as they could be different depending on the mode). 

So the struct would be:

```julia
struct GRUCell{M, Ai, Ah, V, S}
  Wi::Ai
  Wh::Ah
  b::V
  state0::S
end
```

Where M is the mode. My instinct is to put this first to make dispatch very clear, but we could put it elsewhere.

**Comment by darsnack** [comment]
Repeating what I mentioned from the other thread: `GRU(...; mode = :v3)`/`GRU{:v3}(...)` are just as verbose as `GRUv3(...)`. At first glance, the forward passes seem quite different for [the line @mkschleg highlighted](https://github.com/FluxML/Flux.jl/blob/2cf128816e73a704e31bc420cfaf57fdb017b0d1/src/layers/recurrent.jl#L257). Different forward pass == different types makes sense to me here. Having conditional forward passes based on a parameter is messier IMO, and I don't there is a function argument we can pass in that runs the two alternate passes (w/o this function being complex).

We can still reduce code duplication. As an example:
```julia
function _gru_output(Wi, Wh, x, h)
   gx, gh = Wi*x, Wh*h
   r = σ.(gate(gx, o, 1) .+ gate(gh, o, 1) .+ gate(b, o, 1))
   z = σ.(gate(gx, o, 2) .+ gate(gh, o, 2) .+ gate(b, o, 2))

   return r, z
end
```
Then use `_gru_output` in all the variants. IMO it makes sense to reduce duplication via the API when it results in a intuitive interpretation of what's going on under the hood. I don't think that's the case here.

File: src/layers/recurrent.jl (hunk 1 above)
Comment (ToucheSir) [reply src/layers/recurrent.jl:248]: Worth pointing out that TF and JAX-based libraries do allow you to customize the activation. I presume PyTorch doesn't because it lacks a non-CuDNN path for it's GPU RNN backend.

Comment (darsnack) [reply src/layers/recurrent.jl:248]: Interesting, is it just the output activation or all of them?

Comment (mkschleg) [reply src/layers/recurrent.jl:248]: All of them I believe: [tensorflow GRU](https://www.tensorflow.org/api_docs/python/tf/keras/layers/GRU)

Comment (ToucheSir) [reply src/layers/recurrent.jl:248]: All of them, IIUC. There's a distinction between "activation" functions (by default `tanh`) and "gate" functions (by default `sigmoid`).

Comment (ToucheSir) [reply src/layers/recurrent.jl:248]: For a non-Google implementation, here's [MXNet](https://mxnet.apache.org/versions/1.8.0/api/python/docs/api/gluon/rnn/index.html#mxnet.gluon.rnn.LSTMCell).

**Comment by ToucheSir** [comment]
Exploring the other extreme for a second, I wonder if we could make `GRUv3Cell` the default GRU cell and set `Wh_h̃` to `Zeros` in the `GRUCell` constructor. In theory, this would reduce `.+ (m.Wh_h̃ * (r .* h))` to a no-op and allow us to use one codepath.

**Comment by mkschleg** [comment]
Maybe I'm misunderstanding. The forward for the GRUv3Cell would then look like

```julia
  h̃ = tanh.(gate(gx, o, 3) .+ r .* gate(gh, o, 3) .+ (m.Wh_h̃ * (r .* h)) .+ gate(b, o, 3))
```

How would we turn off ```r .* gate(gh, o, 3)``` for the v1 version?

File: src/layers/recurrent.jl (hunk 1 above)
Comment (ToucheSir) [reply src/layers/recurrent.jl:248]: Prior art in Flux: https://github.com/FluxML/Flux.jl/pull/964

Comment (DhairyaLGandhi) [reply src/layers/recurrent.jl:248]: Yeah, I agree that we should have the activations in Flux generally across all layers.

**Comment by darsnack** [comment]
If I understand it correctly, `Wh` in v1 == `[Wh Wh_h̃]` in v3. So always adopting the `gate(Wh * h, o, 1) .+ Wh_h̃ * (r .* h))` means that `Wh` takes on a slightly different interpretation each time. In v1, `Wh` is plain `Wh` and `Wh_h̃ == Zeros()`. In v3, we would initialize `Wh` to be the first two slices and `Wh_h̃` to be the last.

**Comment by ToucheSir** [comment]
Sorry, that was my mistake! Missed that the v1 also had a middle term and assumed it was another case of https://julialang.zulipchat.com/#narrow/stream/238249-machine-learning/topic/Elman.20RNN.20definition.

File: src/layers/recurrent.jl [hunk 4]
```diff
@@ -233,6 +234,62 @@ function Base.getproperty(m::GRUCell, sym::Symbol)
...
+  b::V
+  Wh_h̃::A
+  state0::S
+end
+
+GRUv3Cell(in, out; init = glorot_uniform, initb = zeros32, init_state = zeros32) =
+    GRUv3Cell(init(out * 3, in), init(out * 2, out), initb(out * 3), 
+              init(out, out), init_state(out,1))
+
+function (m::GRUv3Cell{A,V,<:AbstractMatrix{T}})(h, x::Union{AbstractVecOrMat{T},OneHotArray}) where {A,V,T}
```
Comment (DhairyaLGandhi) [inline src/layers/recurrent.jl:252]: Do we need to remove the types on the input and parameters?

Comment (darsnack) [reply src/layers/recurrent.jl:252]: I believe these were introduced as part of #1521. We should tackle them separately for all recurrent cells.

Comment (mkschleg) [reply src/layers/recurrent.jl:252]: Maybe we should have a central issue which details all the updates to recurrent cells the discussion in this PR and related issue has mentioned?

Comment (mkschleg) [reply src/layers/recurrent.jl:252]: I'm definitely invested in the recurrent architectures for flux, so would like to help. But knowing all the outstanding issues is out of scope for what I can use my time for right now.

Comment (ToucheSir) [reply src/layers/recurrent.jl:252]: Yeah, let's get this through and litigate general changes to the RNN interface in a separate issue.

**Comment by mkschleg** [comment]
What needs to be done to push this PR through?

File: src/layers/recurrent.jl [hunk 5]
```diff
@@ -233,6 +234,62 @@ function Base.getproperty(m::GRUCell, sym::Symbol)
...
+function Base.getproperty(m::GRUv3Cell, sym::Symbol)
+  if sym === :h
+    Zygote.ignore() do
+      @warn "GRUv3Cell field :h has been deprecated. Use m::GRUv3Cell.state0 instead."
+    end
+    return getfield(m, :state0)
+  else
+    return getfield(m, sym)
+  end
+end
```
Comment (darsnack) [inline src/layers/recurrent.jl:291]: Don't need this since `GRUv3Cell.h` never existed previously.

File: src/layers/recurrent.jl [hunk 6]
```diff
@@ -233,6 +234,62 @@ function Base.getproperty(m::GRUCell, sym::Symbol)
...
+
+GRUv3Cell(in, out; init = glorot_uniform, initb = zeros32, init_state = zeros32) =
+    GRUv3Cell(init(out * 3, in), init(out * 2, out), initb(out * 3), 
+              init(out, out), init_state(out,1))
+
+function (m::GRUv3Cell{A,V,<:AbstractMatrix{T}})(h, x::Union{AbstractVecOrMat{T},OneHotArray}) where {A,V,T}
+  b, o = m.b, size(h, 1)
+  gx, gh = m.Wi*x, m.Wh*h
+  r = σ.(gate(gx, o, 1) .+ gate(gh, o, 1) .+ gate(b, o, 1))
+  z = σ.(gate(gx, o, 2) .+ gate(gh, o, 2) .+ gate(b, o, 2))
```
Comment (darsnack) [inline src/layers/recurrent.jl:256]: Separate this out into a helper for code re-use across `GRUCell` and `GRUv3Cell`.

File: src/layers/recurrent.jl [hunk 7]
```diff
@@ -233,6 +234,62 @@ function Base.getproperty(m::GRUCell, sym::Symbol)
...
+
+Base.show(io::IO, l::GRUv3Cell) =
+  print(io, "GRUv3Cell(", size(l.Wi, 2), ", ", size(l.Wi, 1)÷3, ")")
+
+"""
+    GRUv3(in::Integer, out::Integer)
+
+[Gated Recurrent Unit](https://arxiv.org/abs/1406.1078) layer. Behaves like an
+RNN but generally exhibits a longer memory span over sequences. This implements
+the version proposed in v3 of the above arxiv paper.
```
Comment (darsnack) [inline src/layers/recurrent.jl:273]:
```suggestion
the variant proposed in v3 of the referenced paper.
```

File: src/layers/recurrent.jl [hunk 8]
```diff
@@ -233,6 +234,62 @@ function Base.getproperty(m::GRUCell, sym::Symbol)
...
+
+@functor GRUv3Cell
+
+Base.show(io::IO, l::GRUv3Cell) =
+  print(io, "GRUv3Cell(", size(l.Wi, 2), ", ", size(l.Wi, 1)÷3, ")")
+
+"""
+    GRUv3(in::Integer, out::Integer)
+
+[Gated Recurrent Unit](https://arxiv.org/abs/1406.1078) layer. Behaves like an
```
Comment (darsnack) [inline src/layers/recurrent.jl:271]:
```suggestion
[Gated Recurrent Unit](https://arxiv.org/abs/1406.1078v3) layer. Behaves like an
```

File: src/layers/recurrent.jl [hunk 9]
```diff
@@ -213,7 +213,8 @@ Base.show(io::IO, l::GRUCell) =
     GRU(in::Integer, out::Integer)
 
 [Gated Recurrent Unit](https://arxiv.org/abs/1406.1078) layer. Behaves like an
```
Comment (darsnack) [inline src/layers/recurrent.jl:215]:
```suggestion
[Gated Recurrent Unit](https://arxiv.org/abs/1406.1078v1) layer. Behaves like an
```

File: src/layers/recurrent.jl [hunk 10]
```diff
@@ -213,7 +213,8 @@ Base.show(io::IO, l::GRUCell) =
...
-RNN but generally exhibits a longer memory span over sequences.
+RNN but generally exhibits a longer memory span over sequences. This implements
+the cell proposed in v1 of the above paper.
```
Comment (darsnack) [inline src/layers/recurrent.jl:217]:
```suggestion
the variant proposed in v1 of the referenced paper.
```

**Review by darsnack — CHANGES_REQUESTED** [review]

File: src/layers/recurrent.jl (hunk 6 above)
Comment (mkschleg) [reply src/layers/recurrent.jl]: Done in latest commit.

File: src/layers/recurrent.jl (hunk 5 above)
Comment (mkschleg) [reply src/layers/recurrent.jl]: Done.

**Comment by ToucheSir** [comment]
Running CI to see what it thinks. Otherwise I think the only other change would be to squash some of those intermediate update commits.

**Review by darsnack — CHANGES_REQUESTED** [review]
Can you add an entry to NEWS.md?

File: src/layers/recurrent.jl [hunk 11]
```diff
@@ -183,6 +183,15 @@ end
 
 # GRU
 
+function _gru_output(Wi, Wh, b, x, h)
```
Comment (DhairyaLGandhi) [inline src/layers/recurrent.jl:186]: Needs a 2-space indent to match Flux style

**Comment by mkschleg** [comment]
@darsnack Where should I put the announcement in NEWS.md?

**Comment by darsnack** [comment]
As a new bullet under 0.12.7 (we're missing a 0.12.6 new entry for some reason).

**Comment by mkschleg** [comment]
Also, how would I squash those commits? I've not had to do that before.

**Comment by darsnack** [comment]
https://stackoverflow.com/questions/35703556/what-does-it-mean-to-squash-commits-in-git

It's just 3 extra commits than necessary, so I don't think you need to bother. In the future, you can go to the "Files" tab of the PR to add review suggestions to a single commit batch (instead of accepting each suggestion as a separate commit).

**Review by darsnack — APPROVED** [review]
Looks good, just gotta wait for CI to pass.
